test/e2e/data/infrastructure-metal3/cluster-template-upgrade-workload/kustomization.yaml

```
@@ -0,0 +1,5 @@
+resources:
+- ../bases/cluster
+patchesStrategicMerge:
+  - ../cluster-template-ubuntu/controlplane-kubeadm-config.yaml
+  - ../cluster-template-ubuntu/workers-kubeadm-config.yaml
```

This would change to 
```suggestion
resources:
- ../bases/ubuntu-kubeadm-config
```

---

I think both ways have pros and cons instead of having everything in bases I think it is relatively explicit to to have patches in other folders in changes in different files rather then having duplication in bases based on the OS name

---

It depends on how you look at it I guess. But if you don't want more bases we could also solve it by pointing the `cluster-template-ubuntu` to use the `cluster-template-ubuntu-workload` under `resources` instead. IT would have the same effect so we can drop `--load-restrictors`.

Something like this:

```yaml
# cluster-template-ubuntu
resources:
- ../bases/ippool
- ../cluster-template-ubuntu-workload
```

```yaml
# cluster-template-ubuntu
resources:
- ../bases/cluster
patchesStrategicMerge:
  - ./cluster-template-ubuntu/controlplane-kubeadm-config.yaml
  - ./cluster-template-ubuntu/workers-kubeadm-config.yaml
```

---

I tried 
```yaml
# ubuntu
resources:
- ../bases/cluster
- ../bases/ippool
- ../cluster-template-upgrade-workload
patchesStrategicMerge:
  - ./controlplane-kubeadm-config.yaml
  - ./workers-kubeadm-config.yaml
```
```yaml
# cluster-template-upgrade-workload
resources:
- ../bases/cluster
patchesStrategicMerge:
  - ../cluster-template-ubuntu/controlplane-kubeadm-config.yaml
  - ../cluster-template-ubuntu/workers-kubeadm-config.yaml

```
But still failing 
`security; file 'cluster-api-provider-metal3/test/e2e/data/infrastructure-metal3/cluster-template-ubuntu/controlplane-kubeadm-config.yaml' is not in or below 'cluster-api-provider-metal3/test/e2e/data/infrastructure-metal3/cluster-template-upgrade-workload'`

---

The point is you need to move the files so you never have `..` for `patchesStrategicMerge`. So you can move them to `cluster-template-upgrade-workload` and use that as "base" for `cluster-template-ubuntu`.

```
└── infrastructure-metal3
    ├── bases
    │   ├── cluster
    │   └── ippool
    ├── cluster-template-centos
    ├── cluster-template-ubuntu
    └── cluster-template-upgrade-workload
        ├── controlplane-kubeadm-config.yaml
        └── workers-kubeadm-config.yaml
```

```yaml
# cluster-template-upgrade-workload
resources:
- ../bases/cluster
patchesStrategicMerge:
  # NOTE: No double dots here!
  - ./cluster-template-ubuntu/controlplane-kubeadm-config.yaml
  - ./cluster-template-ubuntu/workers-kubeadm-config.yaml
```

Now since cluster-template-upgrade-workload already has everything we need except the IP pool, we just add that in the cluster-template-ubuntu:
```yaml
# cluster-template-ubuntu
resources:
- ../bases/ippool
- ../cluster-template-upgrade-workload
# No patches needed, we just grab the IP pool and the cluster-template-upgrade-workload.
```

---

Semantically upgrade use ubuntu as a base because we can also use upgrade based on centos if we want and then it does not make sense to me it is strange that ubuntu template is based on the upgrade, I prefer that we keep this kustomize folder as it is ?

---

Then please go with my first suggestion of moving them to bases. Referencing patches outside of the folder is very confusing and makes it impossible to reuse the kustomization. This is why I want to get rid of it, and why there is a flag for it: it is not the "normal" way.

Having the kubeadm-config in the bases is quite nice. Then we have one base for the cluster, one for the IP pool and one each for ubuntu and centos specific clusters. Then in the overlays we just pick what we need, e.g. ubuntu + IP pool.

---

done

---

Thanks LGTM now! :slightly_smiling_face: